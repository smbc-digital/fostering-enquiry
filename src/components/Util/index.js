import 'isomorphic-fetch'
import sanitizeHtml from 'sanitize-html'

async function submitForm(context) {
    const convertedContext = {}
    
    Object.keys(context).map(key => {
        if (context[key] !== undefined && context[key].value !== undefined) {
            if(typeof context[key].value === 'object'){
                convertedContext[key] = context[key].value
            } else {
                convertedContext[key] = sanitizeHtml(context[key].value, {
                    allowedTags: [],
                    allowedAttributes: {}
                  })
            }
        }
    })

    var doneConvertedContext = JSON.stringify(convertedContext)
    try {
        const result = await Promise.race([
            await fetch('/contact-the-fostering-team/submit', {
                method: 'POST',
                headers: {
                    'Accept': 'application/json; charset=utf-8',
                    'Content-Type': 'application/json; charset=utf-8'
                },
                body: doneConvertedContext
            }),
            new Promise((resolve, reject) => {
                setTimeout(() => reject(new Error('Timeout')), 10000)
            })
        ])

        const responseObject = await result.text()
        const jsonResponse = JSON.parse(responseObject)
        const paymentUrl = jsonResponse.url
        const caseId = jsonResponse.caseID

        return {
            status: result.status,
            url: paymentUrl,
            caseId: caseId
        }
    }
    catch (error) {
        return {
            status: 400
        }
    }
}

export default submitForm