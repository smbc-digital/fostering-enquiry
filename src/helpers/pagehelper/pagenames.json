[
	{
		"number": 1,
		"route": "/fostering-enquiry/about-yourself"
	},
	{
		"number": 2,
		"route": "/fostering-enquiry/contact-details"
	},
	{
		"number": 3,
		"route": "/fostering-enquiry/submit"
	}
	,
	{
		"number": 4,
		"route": "/fostering-enquiry/success"
	}
	,
	{
		"number": 5,
		"route": "/fostering-enquiry/error"
	}
]