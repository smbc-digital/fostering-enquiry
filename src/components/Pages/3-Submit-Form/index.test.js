import { React, mount} from '../../../helpers/SetupTest'
import { SubmitForm } from './index'
import * as utilSubmitForm from '../../Util'
import { getPageRoute } from '../../../helpers/pagehelper'
import renderer from 'react-test-renderer'

describe('SubmitForm', () => {

    beforeEach(() => {
        fetch.resetMocks()
    })

	it('should call push on submit with the correct page when response is 200', async () => {

        const context = {
            displayRecaptcha: 'false',
            onFormSubmission: jest.fn()
        }
		const history = { push: jest.fn() }
        utilSubmitForm.default = jest.fn().mockReturnValue({ status: 200, caseId: '12345678' })
    
        const wrapper = mount(<SubmitForm context={context} history={history} />)
    
        // Act
        await wrapper.find('form').simulate('submit')
    
        // Assert
        expect(context.onFormSubmission).toBeCalledWith('12345678')
        expect(history.push).toHaveBeenCalledWith(getPageRoute(4))
    }),

    it('should call push on submit with an error page when reponse is not 200', async () => {
        // Arrange
        const context = {
            displayRecaptcha: 'false',
            onFormSubmission: jest.fn()
        }
        const history = { push: jest.fn() }
        utilSubmitForm.default = jest.fn().mockReturnValue({ status: 400, caseId: '' })

        const wrapper = mount(<SubmitForm context={context} history={history} />)
        
       // Act
		await wrapper.find('form').simulate('submit')

        // Assert
        expect(history.push).toHaveBeenCalledWith(getPageRoute(5))
    })

    it('should not call push on submit when recaptcha is invalid', () => {
        // Arrange
        document.body.innerHTML = '<span id="displayRecaptcha" class="hidden">true</span>'
        const data = {
            displayRecaptcha: true
		}

        let history = { push: jest.fn() }
        let wrapper = mount(<SubmitForm context={data} history={history} />)

        // Act
        wrapper.find('button').simulate('click')

        // Assert
        expect(history.push).not.toBeCalled()
    })

    it('should enable submit button when recaptcha is valid',() => { 

        document.body.innerHTML = '<span id="displayRecaptcha" class="hidden">true</span>'
        // Arrange
		const context = {
            displayRecaptcha: true
        }
        const history = {
            push: jest.fn()
        }

        // Act
        const wrapper = mount(<SubmitForm context={context} history={history} />)

        wrapper.setState({ recaptchaValid: true })

        // Assert
        expect(wrapper.find('button').props().disabled).toBeFalsy()
    })

    it('should enable submit button when displayRecaptcha is false',() => { 

        document.body.innerHTML = '<span id="displayRecaptcha" class="hidden">true</span>'
        // Arrange
		const context = {
            displayRecaptcha: false
        }
        const history = {
            push: jest.fn()
        }

        // Act
        const wrapper = mount(<SubmitForm context={context} history={history} />)

        wrapper.setState({ recaptchaValid: false })

        // Assert
        expect(wrapper.find('button').props().disabled).toBeFalsy()
    })

    it('should set recaptchaValid in state to true', () => {
        // Arrange
        document.body.innerHTML = '<span id="displayRecaptcha" class="hidden">true</span>'
        const data = {
            displayRecaptcha: false,
		}

        let history = { push: jest.fn() }
        let wrapper = mount(<SubmitForm context={data} history={history} />)

        // Act
        wrapper.instance().onChangeRecaptcha('testRecaptchaValue')

        // Assert
        expect(wrapper.state().recaptchaValid).toBe(true)
    })

    it('should set recaptchaValid in state to false', () => {
        // Arrange
        document.body.innerHTML = '<span id="displayRecaptcha" class="hidden">true</span>'
        const data = {
            displayRecaptcha: true,
		}

        let history = { push: jest.fn() }
        let wrapper = mount(<SubmitForm context={data} history={history} />)

        // Act
        wrapper.instance().onChangeRecaptcha(undefined)

        // Assert
        expect(wrapper.state().recaptchaValid).toBe(false)
    })

    describe('snapshot', () => {
        it('renders correctly', () => {
            const context = {
                recaptchaValid: false,
                isLoading: false
            }
            const history = {
                push: jest.fn()
            }
    
            const tree = renderer
                .create(<SubmitForm context={context} history={history} />)
                .toJSON()
            expect(tree).toMatchSnapshot()
        })
    })
})
