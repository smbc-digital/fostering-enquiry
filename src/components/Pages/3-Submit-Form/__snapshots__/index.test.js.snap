// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubmitForm snapshot renders correctly 1`] = `
<form
  onSubmit={[Function]}
>
  <h1 />
  <h2>
    Submit your form
  </h2>
  <p>
    Now that you've completed the form, use the button below to submit it.
  </p>
  <button
    className="button-primary"
    disabled={false}
    id="submit-your-form"
    onClick={undefined}
    type="submit"
  >
    Submit
  </button>
  <p>
    <a
      href="#"
      onClick={[Function]}
    >
      Previous
    </a>
  </p>
</form>
`;
